Python/Product/Analysis/Interpreter/SentinelModule.cs

```diff
@@ -34,38 +34,39 @@ sealed class SentinelModule : IPythonModule, IDisposable {
             }
         }
 
-        public IPythonModule WaitForImport(int millisecondsTimeout) {
+        public Task<IPythonModule> WaitForImportAsync(CancellationToken cancel) {
             var mod = _realModule;
             if (mod != null) {
-                return mod;
+                return Task.FromResult(mod);
             }
             if (_thread == Thread.CurrentThread) {
-                return this;
+                return Task.FromResult<IPythonModule>(this);
             }
 
-            var evt = _event;
-            if (evt == null) {
-                evt = new ManualResetEventSlim();
-                evt = Interlocked.CompareExchange(ref _event, evt, null) ?? evt;
+            var tcs = _tcs;
+            if (tcs == null) {
+                tcs = new TaskCompletionSource<IPythonModule>();
+                tcs = Interlocked.CompareExchange(ref _tcs, tcs, null) ?? tcs;
             }
 
-            if (!evt.Wait(millisecondsTimeout)) {
-                return _realModule;
+            if (cancel.CanBeCanceled) {
```

Comment:
I don't understand this code piece. Why `CancellationToken` can't cancel `tcs` directly?

Reply:
If we have multiple waiters and one cancels, that would prevents all the others from receiving the result.

Reply:
Ok, I need to clarify the lifetime of this thing:
- All this code matters only if `importing == true`. When it is `false`, `_realModule == this` and we don't do anything.
- In this mode, `Completes` should be called for every instance of the `SentinelModule` only once inside `TryImportModule` (should we check for `null` there?). 
- `SentinelModule.Dispose` when `importing == true` is called only after `Complete`

If I'm correct, single instance of `SemaphoreSlim` should be enough:

```cs
private readonly SemaphoreSlim _semaphore;
```
```cs
        public SentinelModule(string name, bool importing) {
            _thread = Thread.CurrentThread;
            Name = name;
            if (importing) {
                _semaphore = new SemaphoreSlim(0, 1);
            } else {
                _realModule = this;
            }
        }
```
```cs
        public Task<IPythonModule> WaitForImport(CancellationToken cancellationToken) {
            var mod = _realModule;
            if (mod != null) {
                return Task.FromResult(mod);
            }
            if (_thread == Thread.CurrentThread) {
                return Task.FromResult(this);
            }

            var task = _semaphore.WaitAsync(cancellationToken);
            return _realModule != null ? Task.FromResult(_realModule): 
        }
```
```cs
public void Complete(IPythonModule module) {
    _realModule = module;
    while (sm.CurrentCount == 0) {
        sm.Release();     
    }
}
```

Reply:
Sure, that'll work too. It's basically doing the same thing :)